cron.py

```diff
@@ -757,6 +763,106 @@ def delete_old_monitoring():
     LOG.d("delete monitoring records older than %s, nb row %s", max_time, nb_row)
 
 
+def _hibp_check(api_key, queue):
+    """
+    Uses a single API key to check the queue as fast as possible.
+
+    This function to be ran simultaneously (multiple _hibp_check functions with different keys on the same queue) to make maximum use of multiple API keys.
+    """
+    while True:
+        try:
+            alias_id = queue.get_nowait()
+        except Empty:
+            return
+
+        alias = Alias.get(alias_id)
+        LOG.d("Checking HIBP for %s", alias)
+
+        request_headers = {
+            "user-agent": "SimpleLogin",
+            "hibp-api-key": api_key,
+        }
+        r = requests.get(
+            f"https://haveibeenpwned.com/api/v3/breachedaccount/{urllib.parse.quote(alias.email)}",
+            headers=request_headers,
+        )
+
+        if r.status_code == 200:
+            # Breaches found
+            alias.hibp_breaches = [Hibp.get_by(id=entry["Name"]) for entry in r.json()]
+        elif r.status_code == 404:
+            # No breaches found
+            alias.hibp_breaches = []
+        else:
+            LOG.error(
+                "An error occured while checking alias %s: %s - %s",
+                alias,
+                r.status_code,
+                r.text,
+            )
+            return
+
+        db.session.add(alias)
+        db.session.commit()
+
+        LOG.d("Updated breaches info for %s", alias)
+
+        sleep(1.5)
+
+
+def check_hibp():
+    """
+    Check all aliases on the HIBP (Have I Been Pwned) API
+    """
+    LOG.d("Checking HIBP API for aliases in breaches")
+
+    if len(HIBP_API_KEYS) == 0:
+        LOG.exception("No HIBP API keys")
+        return
+
+    LOG.d("Updating list of known breaches")
+    r = requests.get("https://haveibeenpwned.com/api/v3/breaches")
+    for entry in r.json():
+        Hibp.get_or_create(id=entry["Name"])
+
+    db.session.commit()
+    LOG.d("Updated list of known breaches")
+
+    LOG.d("Preparing list of aliases to check")
+    queue = multiprocessing.Queue()
+    for alias in Alias.query.order_by(Alias.hibp_last_check.asc().nullsfirst()).all():
```

Comment:
I wonder if this query can be sped up by using something like

```
max_date = arrow.now().shift(days=-7)
for alias in Alias.query.filter(or_(Alias.hibp_last_check.is_(None), Alias.hibp_last_check < max_date))).all():
    queue.put(alias.id)
```

so we don't have to evaluate `alias.needs_hibp_scan()` for every alias which can take some times for +500K aliases :).

Reply:
Will also need the do the order but yeah, databases tend to be way faster at this. I'll change that.

Reply:
Fixed in 73c2bece704567872cf3067bc8b4df7435535ef0.